### PR TITLE
Fix the Fusion Pass

### PR DESCRIPTION
**Pull request by zdevito**
* Update fusion pass to work with new op format.
* Add a test for Concat optimization.
* Change tracer to generate single output nodes in the simple case.
* Update the ONNX broadcast logic to work with single output nodes.

### CODE REVIEW

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 1]
```diff
@@ -515,9 +515,6 @@ struct StageClosure {
     IR_ELSEIF(Select)
       // No-op. Selects are handled by their inputs.
       return nullptr;
-#define IR_ELSEIF_TRIVIAL(NAME, FNAME) IR_ELSEIF(NAME) return std::make_shared<FNAME>();
-    IR_ELSEIF_TRIVIAL(Add, Add)
-    IR_ELSEIF_TRIVIAL(Mul, Mul)
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: you forgot to remove `#undef` below

File: torch/csrc/jit/export.cpp [hunk 2]
```diff
@@ -290,17 +289,10 @@ void fuseBroadcast(const std::shared_ptr<Graph>& graph) {
                          rhs->type()->expect<TensorType>()->sizes()) // to
        ) continue;
 
-    auto *new_n = graph->createClone(n, [&](Node* n) { return n == rhs ? new_rhs : n; });
-    new_n->i_(kbroadcast, 1);
-    new_n->insertAfter(n);
-    n->replaceAllUsesWith(new_n);
-    it.destroyCurrent();
+    n->replaceInput(n->inputs().size() - 1, new_rhs);
+    n->i_(kbroadcast,1);
```
Comment (apaszke) [inline torch/csrc/jit/export.cpp]: wooow, nice change

File: torch/csrc/jit/interned_strings.h [hunk 3]
```diff
@@ -23,15 +24,17 @@ _(Sub) \
 _(Pow) \
 _(Sigmoid) \
 _(Tanh) \
+_(mul) \
```
Comment (apaszke) [inline torch/csrc/jit/interned_strings.h]: We should be able to remove a lot of these upper case strings now

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 4]
```diff
@@ -41,11 +40,15 @@ struct GraphFuser {
...
+  bool hasSpecialAlpha(Node * node) {
+    if(!node->hasAttribute(kalpha))
+      return false;
+    return at::Scalar(node->t(kalpha)).toDouble() != 1;
+  }
   bool isFusable(Node * node) {
     if (!node->hasType()) return false;
     if (node->kind() == kFusionGroup) return true;
-    return isSimpleMap(node) && isCuda(node);
+    return isSimpleMap(node) && !hasSpecialAlpha(node) && isCuda(node);
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: Can you leave a TODO for implementing special alpha? It seems quite easy to implement, as far as I understand you just didn't want to spend time on it right now

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 5]
```diff
@@ -192,12 +195,7 @@ struct GraphFuser {
   }
 
   bool isChunk(Node * node) {
-    if (node->kind() != kSplit) return false;
-    // All splits have to be equal
-    auto & splits = node->is(ksplit);
-    for (auto s : splits)
-      if (s != splits[0]) return false;
-    return true;
+    return (node->kind() == ksplit);
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: nit: unnecessary parenthesis

File: torch/csrc/jit/passes/peephole.cpp [hunk 6]
```diff
@@ -15,11 +15,7 @@ void PeepholeOptimize(std::shared_ptr<Graph>& graph) {
 
     if (n->kind() == kexpand) {
       if (n->is(ksize) == n->input()->type()->expect<TensorType>()->sizes()) {
-        // Attractive but wrong way to do this:
-        // n->replaceAllUsesWith(n->input());
-        auto* out = n->outputs().at(0);
-        out->replaceAllUsesWith(n->input());
-        out->destroy();
+        n->replaceAllUsesWith(n->input());
```
Comment (apaszke) [inline torch/csrc/jit/passes/peephole.cpp]: Why is it still attractive, but no longer wrong?

File: torch/csrc/jit/tracer.cpp [hunk 7]
```diff
@@ -106,15 +106,20 @@ Node* recordTrace(std::string op, // TODO: make this a Symbol
...
+      setValueTrace(state, output, value);
+    }
+  };
+  if(outputs.size() == 1) {
+    assignOutput(outputs[0],n);
+  } else {
+    int i = 0;
+    for (Variable output : outputs) {
+      assignOutput(output, graph->appendNode(graph->createSelect(n, i)));
+      i++;
```
Comment (apaszke) [inline torch/csrc/jit/tracer.cpp]: Can you leave this for a separate commit? I think there were some places in Ed's latest commits (now in master), that assumed we never have single-return ops

Comment (apaszke) [reply torch/csrc/jit/tracer.cpp]: also nit: use a for loop if you need the index

**Review by apaszke — COMMENTED** [review]
Looks good for the most part. @ezyang should comment on how multireturn-node-only his changes were

**Comment by apaszke** [comment]
Also, do you know why so many expects changed unique numbers?

**Comment by ezyang** [comment]
It's because, if you're not allocating select nodes, the numberings all change, because the select nodes use up unique numbers.

File: torch/csrc/jit/passes/peephole.cpp (hunk 6 above)
Comment (ezyang) [reply torch/csrc/jit/passes/peephole.cpp]: Because the expand node no longer has selects, so you can replace it directly.

File: torch/csrc/jit/export.cpp (hunk 2 above)
Comment (ezyang) [reply torch/csrc/jit/export.cpp]: Yeah, I did this in my PR as well :)

File: torch/csrc/jit/interned_strings.h (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/interned_strings.h]: They still happen in ONNX, so a lot of them need to remain.

Comment (apaszke) [reply torch/csrc/jit/interned_strings.h]: Of course they can happen, but do we ever use symbol constants to refer to them?

File: torch/csrc/jit/tracer.cpp (hunk 7 above)
Comment (zdevito) [reply torch/csrc/jit/tracer.cpp]: We can't leave it out, it would require extensive changes in the fusion passes and the fusion compiler that we would just need to change back. I fixed _most_ of the onnx issues, I believe. Moving the ONNX passes out of the repo prevents fixing all of them.

File: torch/csrc/jit/interned_strings.h (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/interned_strings.h]: jit::kAdd still happens. I didn't bother to comb through the others.

Comment (ezyang) [reply torch/csrc/jit/interned_strings.h]: Not for this PR, but since the ATen ops are generated by cwrap, we should autogen their interned strings as well.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 8]
```diff
@@ -131,13 +131,17 @@ std::string nodeName(Node * n) {
...
-  // TODO: support both float and int constants
-  {kAddConstant,     [](Node* n) { return std::to_string(n->f(kvalue)) + " + ${0}"; }},
+  {ksigmoid,         [](Node*) { return "1.f / (1.f + expf(-${0}))"; }},
+  {ktanh,            [](Node*) { return "tanhf(${0})"; }},
+  {kmul,             [](Node*) { return "${0} * ${1}"; }},
+  {kadd,             [](Node*n) -> std::string {
+    if(n->inputs().size() == 2)
+      return "${0} + ${1}";
+    else // TODO: support both float and int constants
+      return "${0} + " + std::to_string(n->f(kvalue));
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Are you sure this works? IIRC, `value` is recorded as a scalar `z` not a float `f`.

File: torch/csrc/jit/tracer.cpp [hunk 9]
```diff
@@ -106,15 +106,18 @@ Node* recordTrace(std::string op, // TODO: make this a Symbol
...
+  if(outputs.size() == 1) {
+    assignOutput(outputs[0],n);
+  } else {
+    for(size_t i = 0; i < outputs.size(); i++) {
+      assignOutput(outputs[i], graph->appendNode(graph->createSelect(n, i)));
     }
-    i++;
   }
 
   /*
```
Comment (ezyang) [inline torch/csrc/jit/tracer.cpp]: Oy, don't forget to delete this comment

**Comment by ezyang** [comment]
This is going to break ONNX but I can probably try to harmonize it.

**Comment by zdevito** [comment]
I tested it against the pytorch-onnx repo, using one of your big update patches. It seemed mostly working.

**Comment by zdevito** [comment]
I updated this to fix the addconstant case

**Comment by zdevito** [comment]
Test failures seem spurious now, so this should be ready.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 10]
```diff
@@ -416,6 +416,11 @@ void CompiledFusionFunction::launch(uint32_t numel, void ** arguments) {
   //std::cout << "maxBlocks = " << maxBlocks << " needed blocks: " << ceilDiv(numel,blockSize)
   //          << " numblocks =  " << numBlocks;
 
+  // it is possible that this is the first cuda call on this thread
+  // so make sure we initialize the Driver API's context
+  // cudaFree(0) accomplishes this.
+  cudaFree(0);
```
Comment (apaszke) [inline torch/csrc/jit/fusion_compiler.cpp]: We have to check that this is not a GPU sync point before this is merged

**Comment by ezyang** [comment]
Commandeering this patch.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 10 above)
Comment (ezyang) [reply torch/csrc/jit/fusion_compiler.cpp]: According to https://stackoverflow.com/questions/12539593/is-cudafree-asynchronous cudaFree is synchronous.

Comment (zdevito) [reply torch/csrc/jit/fusion_compiler.cpp]: While cudaFree can sometimes block, cudaFree(0) doesn't do anything but initialize the context and (unfortunately) has become the idiomatic way to do this: https://devtalk.nvidia.com/default/topic/911817/cudamalloc-is-too-expensive-and-gpu-memories/?offset=6
